clippy_lints/src/unused_box.rs

```diff
@@ -0,0 +1,79 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, ty::implements_trait};
+use rustc_errors::Applicability;
+use rustc_hir::{intravisit::FnKind, Body, FnDecl, FnRetTy, HirId};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+use rustc_typeck::hir_ty_to_ty;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Checks for a return type containing a `Box<T>` where `T` implements `Sized`
+    ///
+    /// ### Why is this bad?
+    ///
+    /// It's better to just return `T` in these cases
+    ///
+    /// ### Example
+    /// ```rust
+    /// fn foo() -> Box<String> {
+    ///     Box::new(String::from("Hello, world!"))
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn foo() -> String {
+    ///     String::from("Hello, world!")
+    /// }
+    /// ```
+    #[clippy::version = "1.64.0"]
+    pub UNUSED_BOX,
+    nursery,
+    "Needlessly returning a Box"
+}
+declare_lint_pass!(UnusedBox => [UNUSED_BOX]);
+
+impl LateLintPass<'_> for UnusedBox {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'_>,
+        fn_kind: FnKind<'_>,
+        decl: &FnDecl<'_>,
+        _: &Body<'_>,
+        _: Span,
+        _: HirId,
+    ) {
+        // it's unclear what part of a closure you would span, so for now it's ignored
+        // if this is changed, please also make sure not to call `hir_ty_to_ty` below
+        if let FnKind::Closure = fn_kind {
```

Comment:
we can avoid the Yoda condition by using `matches!` for this instead of using `if let`

```suggestion
        if matches!(fn_kind, FnKind::Closure) {
```

This might be a good improvement for `equatable_if_let` lint :thinking:

Reply:
Just made that change